src/main/java/rx/internal/operators/OperatorBufferPredicateBoundary.java

```diff
@@ -0,0 +1,487 @@
+            error = e;
+            done = true;
+            drain();
+        }
+        
+        @Override
+        public void onCompleted() {
+            done = true;
+            drain();
+        }
+        
+        void requestMore(long n) {
+            if (n > 0) {
+                BackpressureUtils.getAndAddRequest(requested, n);
+                drain();
+            } else
+            if (n < 0) {
+                throw new IllegalArgumentException("n >= 0 required but it was " + n);
+            }
+        }
+        
+        void drain() {
+            if (wip.getAndIncrement() != 0) {
+                return;
+            }
+            
+            final Subscriber<? super List<T>> localSubscriber = actual;
+            final Queue<Object> localQueue = queue;
+            int missed = 1;
+            
+            for (;;) {
+
+                long localRequested = requested.get();
+                long localEmission = 0L;
+                long localConsumption = 0L;
+                List<T> localBuffer = buffer;
+                
+                while (localEmission != localRequested) {
+                    if (localSubscriber.isUnsubscribed()) {
+                        return;
+                    }
+                    
+                    boolean mainDone = done;
+                    
+                    if (mainDone) {
+                        Throwable exception = error;
+                        if (exception != null) {
+                            buffer = null;
+                            localSubscriber.onError(exception);
+                            return;
+                        }
+                    }
+                    
+                    Object notification = localQueue.poll();
+                    boolean empty = notification == null;
+                    
+                    if (mainDone && empty) {
+                        buffer = null;
+                        if (!localBuffer.isEmpty()) {
+                            localSubscriber.onNext(localBuffer);
+                        }
+                        localSubscriber.onCompleted();
+                        return;
+                    }
+                    if (empty) {
+                        break;
+                    }
+                    
+                    T value = nl.getValue(notification);
+                    
+                    localBuffer.add(value);
+                    localConsumption++;
+                    
+                    boolean emit;
+                    
+                    try {
+                        emit = predicate.call(value);
+                    } catch (Throwable ex) {
+                        unsubscribe();
+                        buffer = null;
+                        Exceptions.throwOrReport(ex, localSubscriber, value);
+                        return;
+                    }
+                    
+                    if (emit) {
+                        localSubscriber.onNext(localBuffer);
```

Comment:
As predicate.call could take a non-trivial time we could check for unsubscription before emitting. It's a trade-off between checking the volatile unsubscribed flag and reducing work after unsubscription but in other operators we normally do the unsubscribed check.